internal/driver/incominglistener.go

```diff
@@ -8,60 +8,19 @@ package driver
 
 import (
 	"encoding/json"
-	"fmt"
-	"net/url"
-	"strings"
-	"time"
 
-	mqtt "github.com/eclipse/paho.mqtt.golang"
 	"github.com/edgexfoundry/device-sdk-go/v2/pkg/models"
 	"github.com/edgexfoundry/device-sdk-go/v2/pkg/service"
+
+	"github.com/eclipse/paho.mqtt.golang"
 )
 
 func startIncomingListening() error {
-	var scheme = driver.serviceConfig.MQTTBrokerInfo.IncomingSchema
-	var brokerUrl = driver.serviceConfig.MQTTBrokerInfo.IncomingHost
-	var brokerPort = driver.serviceConfig.MQTTBrokerInfo.IncomingPort
-	var authMode = driver.serviceConfig.MQTTBrokerInfo.IncomingAuthMode
-	var secretPath = driver.serviceConfig.MQTTBrokerInfo.IncomingCredentialsPath
-	var mqttClientId = driver.serviceConfig.MQTTBrokerInfo.IncomingClientId
-	var qos = byte(driver.serviceConfig.MQTTBrokerInfo.IncomingQos)
-	var keepAlive = driver.serviceConfig.MQTTBrokerInfo.IncomingKeepAlive
+	var qos = byte(driver.serviceConfig.MQTTBrokerInfo.Qos)
 	var topic = driver.serviceConfig.MQTTBrokerInfo.IncomingTopic
-
-	uri := &url.URL{
-		Scheme: strings.ToLower(scheme),
-		Host:   fmt.Sprintf("%s:%d", brokerUrl, brokerPort),
-	}
-
-	err := SetCredentials(uri, "Incoming", authMode, secretPath)
-	if err != nil {
-		return err
-	}
-
-	var client mqtt.Client
-	for i := 1; i <= driver.serviceConfig.MQTTBrokerInfo.ConnEstablishingRetry; i++ {
-		client, err = createClient(mqttClientId, uri, keepAlive)
-		if err != nil && i == driver.serviceConfig.MQTTBrokerInfo.ConnEstablishingRetry {
-			return err
-		} else if err != nil {
-			driver.Logger.Error(fmt.Sprintf("Fail to initial conn for incoming data, %v ", err))
-			time.Sleep(time.Duration(driver.serviceConfig.MQTTBrokerInfo.ConnEstablishingRetry) * time.Second)
-			driver.Logger.Warn("Retry to initial conn for incoming data")
-			continue
-		}
-		break
-	}
-
-	defer func() {
-		if client.IsConnected() {
-			client.Disconnect(5000)
-		}
-	}()
-
-	token := client.Subscribe(topic, qos, onIncomingDataReceived)
+	token := driver.mqttClient.Subscribe(topic, qos, onIncomingDataReceived)
```

Comment:
Move subscribe to OnConnect handler so it re-subscribes on reconnects

Reply:
Modified.